ouroboros-network/src/Ouroboros/Network/NodeToNode.hs

```diff
@@ -538,3 +548,74 @@ localNetworkErrorPolicy = ErrorPolicies {
+                       -> m ()
+simpleResponderControl rspFn = forever $
+    -- Fetch, but ignore the result of any miniprotocol so that it can
+    -- be restarted incase the initiator decides to talk to us again.
+    void $ atomically $ fetchResponderResult ChainSyncWithHeadersPtcl
+               `orElse` fetchResponderResult BlockFetchPtcl
+               `orElse` fetchResponderResult TxSubmissionPtcl
+  where
+    fetchResponderResult :: MonadSTM m => NodeToNodeProtocols -> STM m a
+    fetchResponderResult ptcl =
+        let (MiniProtocolResponderControl action) = rspFn ptcl in
+        action
+
+-- | Result of waiting on a list of miniprotocols to finish. Only used by simpleInitiatorControl
+data MiniProtocolClientResult = MpcRestart NodeToNodeProtocols
+                              | MpcResult  NodeToNodeProtocols
+
+-- | Simple initiator control function that starts all NodeToNodeProtocols and if one of them
+-- finishes it will wait for restartDelay seconds before restarting the protocol.
+simpleInitiatorControl :: forall m a.
+                          ( MonadAsync m
+                          , MonadSTM m
+                          , MonadTimer m
+                          )
+                       => DiffTime
+                       -> (NodeToNodeProtocols -> MiniProtocolInitiatorControl m a)
+                       -> m ()
+simpleInitiatorControl restartDelay ctrlFn = do
+    startQ <- atomically $ newTBQueue 3
+    atomically $ do
+        writeTBQueue startQ ChainSyncWithHeadersPtcl
+        writeTBQueue startQ BlockFetchPtcl
+        writeTBQueue startQ TxSubmissionPtcl
+    clientK startQ []
+
+  where
+    clientK :: TBQueue m NodeToNodeProtocols -> [(NodeToNodeProtocols, STM m a)] -> m ()
+    clientK restartQ resultActions = do
+        cr <- atomically $ waitOnRestartQueue restartQ
+                `orElse` foldr (orElse . waitOnMiniProtocolClientResult) retry resultActions
+        case cr of
+                (MpcRestart ptcl) -> do
+                    let (MiniProtocolInitiatorControl restart) = ctrlFn ptcl
+                    resultAction <- atomically restart
+                    clientK restartQ ((ptcl, resultAction):resultActions)
+                (MpcResult ptcl) -> do
+                    -- The return type is ignored, but one could act on the result here
+
+                    -- Schedule a restart of the miniprotocol in the future
+                    void $ async $ delayRestart restartQ ptcl
+                    clientK restartQ $ filter (\(mp, _) -> mp /= ptcl) resultActions
+
+    delayRestart :: TBQueue m NodeToNodeProtocols -> NodeToNodeProtocols -> m ()
+    delayRestart q ptcl = do
+        threadDelay restartDelay
+        atomically $ writeTBQueue q ptcl
+
+    waitOnRestartQueue :: TBQueue m NodeToNodeProtocols -> STM m MiniProtocolClientResult
+    waitOnRestartQueue q = do
+        ptcl <- readTBQueue q
+        return $ MpcRestart ptcl
```

Comment:
This could just be written as `MpcRestart <$> readTBQueue q`.

Reply:
Will address.